pywren/wait.py

```diff
@@ -79,12 +110,57 @@ def _wait(fs, THREADPOOL_SIZE):
     # get the list of all objects in this callset
     callset_id = present_callsets.pop() # FIXME assume only one
 
-    storage_config = wrenconfig.extract_storage_config(wrenconfig.default())
-    storage_handler = storage.Storage(storage_config)
-    callids_done = storage_handler.get_callset_status(callset_id)
+    # note this returns everything done, so we have to figure out
+    # the intersection of those that are done
+    callids_done_in_callset = set(storage_handler.get_callset_status(callset_id))
+
+    not_done_call_ids = set([f.call_id for f in not_done_futures])
+
+    done_call_ids = not_done_call_ids.intersection(callids_done_in_callset)
+    not_done_call_ids = not_done_call_ids - done_call_ids
+
+    still_not_done_futures = [f for f in not_done_futures if (f.call_id in not_done_call_ids)]
 
-    callids_done = set(callids_done)
+    def fetch_future_status(f):
+        return storage_handler.get_call_status(f.callset_id, f.call_id)
 
+
+    pool = ThreadPool(THREADPOOL_SIZE)
+
+    # now try up to max_direct_query_n direct status queries, quitting once
+    # we have return_n done.
+    query_count = 0
+    max_queries = min(max_direct_query_n, len(still_not_done_futures))
+
+    if random_query:
+        random.shuffle(still_not_done_futures)
+
+    while query_count < max_queries:
+
+        if len(done_call_ids) >= return_early_n:
+            break
+        num_to_query_at_once = THREADPOOL_SIZE
+        fs_to_query = still_not_done_futures[query_count:query_count + num_to_query_at_once]
```

Comment:
I think `query_count + num_to_query_at_once` could go out of bound?

Reply:
Python list slices always return as much as they can:
```
In [3]: x = list(range(10))

In [4]: x
Out[4]: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [5]: x[5:15]
Out[5]: [5, 6, 7, 8, 9]
```

so I think this should be ok?